iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java

```diff
@@ -307,6 +366,32 @@ public void close()
 
             log.debug("Closing MQTT connection");
 
+            // When connecting to E4K, the client is expected to publish a message before gracefully closing the connection
+            // so that E4K knows to end the upstream connection for this device. Note that the MQTT connection's "will"
+            // message isn't sufficient to handle this case because the "will" message is only sent on ungraceful disconnections.
+            if (this.config.isConnectingToMqttGateway())
+            {
+                try
+                {
+                    IMqttDeliveryToken token = this.mqttAsyncClient.publish(
+                        E4K_CONNECTION_CLOSE_TOPIC,
+                        E4K_CONNECTION_CLOSE_MESSAGE_PAYLOAD,
+                        E4K_CONNECTION_CLOSE_MESSAGE_QOS,
+                        E4K_CONNECTION_CLOSE_MESSAGE_RETAINED);
+
+                    // If the message fails to be sent/acknowledged in this time span, an MqttException will be thrown here
+                    token.waitForCompletion(E4K_CONNECTION_CLOSE_MESSAGE_TIMEOUT_MILLISECONDS);
+                }
+                catch (MqttException e)
+                {
+                    // This is a bit of an odd case, but it seems likely that this catch block on executes if the connection
+                    // was ungracefully dropped while sending this message. In that case, the "will" message will be read
```

Comment:
There could be an exception associated with the "graceful close" send that wouldn't be accompanied by a network disconnect, right? We'd want to retry in that case, right?

Reply:
Possibly. I want to talk this point through a bit more with the E4K team. It boils down to the question "Does E4K have any kind of failsafe for if the "connection closed" message is somehow lost/ignored"

For now, I'd rather not retry because we already don't have retry logic for ```close``` operations.